op-node/rollup/derive/batch_queue_test.go

```diff
@@ -0,0 +1,323 @@
+package derive
+
+import (
+	"context"
+	"io"
+	"math/rand"
+	"testing"
+
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/testlog"
+	"github.com/ethereum-optimism/optimism/op-node/testutils"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/require"
+)
+
+// fakeBatchQueueOutput fakes the next stage (receive only) for the batch queue
+// It tracks the progress state of the next stage.
+// Upon receiving a batch, relevant characteristic of safeL2Head are immediately advanced.
+type fakeBatchQueueOutput struct {
+	progress   Progress
+	batches    []*BatchData
+	safeL2Head eth.L2BlockRef
+}
+
+var _ BatchQueueOutput = (*fakeBatchQueueOutput)(nil)
+
+func (f *fakeBatchQueueOutput) AddBatch(batch *BatchData) {
+	f.batches = append(f.batches, batch)
+	// Advance SafeL2Head
+	f.safeL2Head.Time = batch.Timestamp
+	f.safeL2Head.L1Origin.Number = uint64(batch.EpochNum)
+}
+
+func (f *fakeBatchQueueOutput) SafeL2Head() eth.L2BlockRef {
+	return f.safeL2Head
+}
+
+func (f *fakeBatchQueueOutput) Progress() Progress {
+	return f.progress
+}
+
+func b(timestamp uint64, epoch eth.L1BlockRef) *BatchData {
+	rng := rand.New(rand.NewSource(1234))
```

Comment:
If we're using the same rng seed for each batch, then they'll all look the same.

Reply:
length is different. Could do fixed length & rng seed based on timestamp